feat: Add S3Exporter for exporting observability data to Amazon S3

mansichauhn opened the pull request:
This PR adds S3Exporter to the aiobs.exporters package, allowing users to export observability data directly to Amazon S3 buckets.

Features:

- Export observability data to Amazon S3 buckets

- Support for AWS authentication (credentials, IAM roles, profiles)

- Configurable bucket, region, and prefix

- Filename templates with placeholders

- Custom metadata support

- from_env() class method for environment variable configuration

- Lazy import of boto3 (optional dependency)

Commit:
feat: Add S3Exporter for exporting observability data to Amazon S3

This PR adds S3Exporter to the aiobs.exporters package, allowing users to export observability data directly to Amazon S3 buckets.

Features:

- Export observability data to Amazon S3 buckets

- Support for AWS authentication (credentials, IAM roles, profiles)

- Configurable bucket, region, and prefix

- Filename templates with placeholders

- Custom metadata support

- from_env() class method for environment variable configuration

- Lazy import of boto3 (optional dependency)

Pranavchiku reviewed (commented):
thanks for working on this @mansichauhn! Let's add tests and then I'll review it.

mansichauhn commented:
looks like #58 is merged.I'll rebase this PR.

Pranavchiku reviewed (commented):
Super work on this @mansichauhn, this is a real good first attempt! I just asked opus 4.5 to review and I feel the following changes are required:

## PR Review: Add S3Exporter

Thanks for the contribution! The implementation looks solid and follows the existing `GCSExporter` patterns well. Here's a review with action items:

---

### ✅ What's Good

- [x] Consistent API design matching `GCSExporter`
- [x] Clear docstrings with usage examples and auth chain documentation
- [x] Lazy `boto3` import with helpful error message
- [x] Proper S3 metadata handling (converts values to strings)
- [x] Flexible authentication (explicit creds + AWS default chain + session tokens)
- [x] Correct prefix normalization logic
- [x] Proper error wrapping with `ExportError`

---

### 📋 Checklist — Required Changes

- [ ] **Add unit tests for `S3Exporter`** in `tests/test_exporters.py`
  - [ ] `test_initialization`
  - [ ] `test_prefix_normalization`
  - [ ] `test_filename_template`
  - [ ] `test_generate_filename`
  - [ ] `test_generate_filename_empty_sessions`
  - [ ] `test_export_success` (with mock boto3 client)
  - [ ] `test_export_upload_content`
  - [ ] `test_export_missing_dependency`
  - [ ] `test_from_env_missing_bucket`
  - [ ] `test_from_env_success`
  - [ ] `test_export_with_custom_metadata`
  - [ ] `test_export_failure`
  - [ ] `test_export_with_filename_override`

- [ ] **Fix `from_env()` to support session token**
  ```python
  def from_env(
      cls,
      bucket_env: str = "AIOBS_S3_BUCKET",
      region_env: str = "AIOBS_S3_REGION",
      prefix_env: str = "AIOBS_S3_PREFIX",
      access_key_env: str = "AWS_ACCESS_KEY_ID",
      secret_key_env: str = "AWS_SECRET_ACCESS_KEY",
      session_token_env: str = "AWS_SESSION_TOKEN",  # <- Add
  ) -> "S3Exporter":
      # ...
      return cls(
          bucket=bucket,
          region=os.getenv(region_env, "us-east-1"),
          prefix=os.getenv(prefix_env, ""),
          aws_access_key_id=os.getenv(access_key_env),
          aws_secret_access_key=os.getenv(secret_key_env),
          aws_session_token=os.getenv(session_token_env),  # <- Add
      )
  ```

---

### 📋 Checklist — Optional / Nice-to-Have

- [ ] Add `s3` optional dependency group to `pyproject.toml`:
  ```toml
  s3 = [
    "boto3>=1.26.0",
  ]
  ```
  And update the `all` group to include it.

- [ ] Add `S3Exporter` import to test file header when adding tests

---

### 💡 Notes

- The test structure can mirror `TestGCSExporter` — see lines 253-439 in `test_exporters.py` for reference
- For mocking, inject the mock client via `exporter._client = mock_client` (same pattern as GCS tests)

Let me know if you have any questions!

Commit:
test: Add unit tests for S3Exporter and fix from_env()

- Add comprehensive unit tests for S3Exporter (mirroring TestGCSExporter)

- Fix from_env() to support AWS_SESSION_TOKEN environment variable

- Add s3 optional dependency group to pyproject.toml

- Add S3Exporter import to test file header

mansichauhn commented:
Hi, just checking if this approach looks fine. 
Meanwhile I’m exploring other issues as well.
@Pranavchiku
